src/main/java/com/mraof/minestuck/item/weapon/MusicPlayerItem.java

```diff
@@ -0,0 +1,194 @@
+	{
+		return entity.getCapability(CapabilityMusicPlaying.MUSIC_PLAYING_CAPABILITY).orElseThrow(() ->
+				new IllegalArgumentException("Expected an music playing for this entity, but " + entity + " does not expose a music playing."));
+	}
+	
+	@Nullable
+	@Override
+	public ICapabilityProvider initCapabilities(ItemStack stack, @Nullable CompoundTag nbt)
+	{
+		return new MusicPlayerItemCapProvider();
+	}
+	
+	public MusicPlayerItem(Builder builder, Properties properties, float volume, float pitch)
+	{
+		super(builder, properties);
+		this.volume = volume;
+		this.pitch = pitch;
+		
+	}
+	
+	public MusicPlayerItem(Builder builder, Properties properties)
+	{
+		super(builder, properties);
+		this.volume = 4.0F;
+		this.pitch = 1.0F;
+	}
+	
+	@Override
+	public InteractionResultHolder<ItemStack> use(Level level, Player playerIn, InteractionHand handIn)
+	{
+		ItemStack musicPlayer = playerIn.getItemInHand(handIn);
+		
+		IItemHandler itemStackHandlerMusicPlayer = getItemStackHandlerMusicPlayer(musicPlayer);
+		IMusicPlaying musicPlayingCap = getMusicPlaying(playerIn);
+		
+		if(!level.isClientSide)
+		{
+			if(playerIn.isCrouching())
+			{
+				Item itemInMusicPlayer = itemStackHandlerMusicPlayer.getStackInSlot(0).getItem();
+				if(itemInMusicPlayer instanceof CassetteItem cassette)
+				{
+					MusicPlayerPacket packet;
+					if(musicPlayingCap.getCassetteType() == EnumCassetteType.NONE)
+					{
+						packet = MusicPlayerPacket.createPacket(playerIn, cassette.cassetteID, volume, pitch);
+						musicPlayingCap.setMusicPlaying(musicPlayer, cassette.cassetteID);
+					} else
+					{
+						packet = MusicPlayerPacket.createPacket(playerIn, EnumCassetteType.NONE, 0, 0);
+						musicPlayingCap.setMusicPlaying(ItemStack.EMPTY, EnumCassetteType.NONE);
+					}
+					MSPacketHandler.sendToTrackingAndSelf(packet, playerIn);
+				}
+			}
+			//open the GUI if right-clicked
+			else
+			{
+				MusicPlayerPacket packet = MusicPlayerPacket.createPacket(playerIn, EnumCassetteType.NONE, 0, 0);
+				musicPlayingCap.setMusicPlaying(ItemStack.EMPTY, EnumCassetteType.NONE);
+				MSPacketHandler.sendToTrackingAndSelf(packet, playerIn); //This will stop the music before opening the GUI
+				
+				NetworkHooks.openGui((ServerPlayer) playerIn, new SimpleMenuProvider((pContainerId, pInventory, pPlayer) ->
+						new CassetteContainerMenu(pContainerId, pInventory, itemStackHandlerMusicPlayer, musicPlayer),
+						new TextComponent("Music Player")));
+			}
+		}
+		return super.use(level, playerIn, handIn);
+	}
+	
+	@Override
+	public void inventoryTick(ItemStack stack, Level level, Entity entityIn, int itemSlot, boolean isSelected)
+	{
+		super.inventoryTick(stack, level, entityIn, itemSlot, isSelected);
+		if(stack.getItem() instanceof MusicPlayerItem)
+		{
+			ItemStack itemInMusicPlayer = getItemStackHandlerMusicPlayer(stack).getStackInSlot(0);
+			stack.getOrCreateTag().putBoolean("HasCassette", itemInMusicPlayer != ItemStack.EMPTY);
+		}
+	}
```

Comment:
a) The instanceof check should be unnecessary since it should be assumed that the item of that stack is this item.
b) Don't use `stack != ItemStack.EMPTY`. Use `!stack.isEmpty()`.
c) Please try overriding `getShareTag()` instead.

Reply:
how would you suggest I use getShareTag() instead of inventoryTick()?

Reply:
`@Nullable
	@Override
	public CompoundTag getShareTag(ItemStack stack)
	{
		CompoundTag tag = stack.getTag();
		if(stack.getItem() instanceof MusicPlayerItem && tag != null)
		{
			ItemStack itemInMusicPlayer = getItemStackHandlerMusicPlayer(stack).getStackInSlot(0);
			tag.putBoolean("HasCassette", itemInMusicPlayer != ItemStack.EMPTY);
			
			return tag;
		}
		
		return super.getShareTag(stack);
	}`

Reply:
I thought about that but that would not help me change the tag according to a change in the container.

Reply:
`getShareTag()` is used if a change is detected in the item stack or in any serialized capability, so this should work to change the tag according to changes in the inventory/item handler. However, the instanceof check should not be needed, and you should return a tag with "HasCassette" set even if the item stack does not have a tag. Just create a new tag in that case. `CompoundTag tag = Objects.requireNonNullElseGet(stack.getTag(), CompoundTag::new);`